### PR TITLE
common-updater-scripts: add allowedVersions parameter

### PR DESCRIPTION
**Pull request by rhendric**
## Description of changes

I [said I'd do it](https://discourse.nixos.org/t/generic-git-updater-allowed-versions/47402), so I did it.

By providing a regex in `allowedVersions`, users of `genericUpdater` or `gitUpdater` can exclude versions that don't match the regex. This can be simpler to express than constructing the complement regex in `ignoredVersions`.

There was no documentation for `ignoredVersions` anywhere, so I didn't add any for this either. Woowoowoowoowoowoo!

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/common-updater/generic-updater.nix [hunk 1]
```diff
@@ -52,7 +54,7 @@ let
 
     function version_is_ignored() {
       local tag="$1"
-      [ -n "$ignored_versions" ] && ${grep} -E "$ignored_versions" <<< "$tag"
+      [ -n "$ignored_versions" ] && ${grep} -E -e "$ignored_versions" <<< "$tag"
```
Comment (jtojnar) [inline pkgs/common-updater/generic-updater.nix:57]: Presumably, this is to “protect a pattern beginning with “-””?

**Review by jtojnar — APPROVED** [review]
Looks alright to me.

**Review by romildo — CHANGES_REQUESTED** [review]
`genericUpdater` is also used by `directoryListingUpdater` and `httpTwoLevelsUpdater`. Please update them so that they become aware of the new `allowedVersions` parameter.

**Comment by rhendric** [comment]
> `genericUpdater` is also used by `directoryListingUpdater` and `httpTwoLevelsUpdater`. Please update them so that they become aware of the new `allowedVersions` parameter.

Oh, huh. I just noticed `directoryListingUpdater` has `extraRegex` which possibly does the same thing, but in Python? Should we deprecate that? Or should I use `extraRegex` as the name of this thing instead?

**Comment by romildo** [comment]
> Oh, huh. I just noticed `directoryListingUpdater` has `extraRegex` which possibly does the same thing, but in Python? Should we deprecate that? Or should I use `extraRegex` as the name of this thing instead?

It was uses the [Beautiful Soup python library](https://pypi.org/project/beautifulsoup4/) to easy scraping information from the web page.

The parameter `extraRegex` was added in https://github.com/NixOS/nixpkgs/pull/206398. It is used only in 4 packages in nixpkgs:

- https://github.com/NixOS/nixpkgs/blob/630d11536201e36502eb89ffa13d9256c2fe99ef/pkgs/applications/misc/eaglemode/default.nix#L61
- https://github.com/NixOS/nixpkgs/blob/d42c8a87674fc1b74975260587efc7266d7cdfcd/pkgs/by-name/lt/ltris/package.nix#L28
- https://github.com/NixOS/nixpkgs/blob/cf6ffffdb145e34593485121609f8a17cc5a884b/pkgs/by-name/lb/lbreakouthd/package.nix#L32
- https://github.com/NixOS/nixpkgs/blob/1dffcb5db1225624677bad74496de09507f6b34f/pkgs/by-name/lp/lpairs2/package.nix#L30

I think `extraRegex` can be deprecated in favor of `allowedVersions`.

@chuangzhu @ehmry @AndersonTorres What do you think about it?

**Comment by rhendric** [comment]
Those are all negative lookaheads; I wonder why they didn't use `ignoredVersions`. @fgaz, am I missing something?

**Comment by AndersonTorres** [comment]
Is there any error message to not mess concurrent usage of both patterns?

**Comment by fgaz** [comment]
@rhendric I don't remember. maybe ignoredVersion didn't exist, or I didn't know about it.

**Review by AndersonTorres — APPROVED** [review]

**Review by romildo — APPROVED** [review]

File: pkgs/games/minetest/default.nix [hunk 2]
```diff
@@ -130,7 +130,8 @@ stdenv.mkDerivation (finalAttrs: {
   '';
 
   passthru.updateScript = gitUpdater {
-    ignoredVersions = "^[^.]+$|.*-android$";
+    allowedVersions = "\\.";
```
[the remark this answers is not in the document]
Comment (rhendric) [reply pkgs/games/minetest/default.nix:133]: Yes, if both `allowedVersions` and `ignoredVersions` are specified, a version must pass both tests to be considered.

**Review by AndersonTorres — APPROVED** [review]
Update the docs, especially about using both parameters, and it's good to go.

**Comment by rhendric** [comment]
@AndersonTorres, what docs? I couldn't find `ignoredVersions`, `genericUpdater`, or `gitUpdater` anywhere in Nixpkgs outside of `.nix` files.

**Comment by AndersonTorres** [comment]
It's even worse than what I've thought about!

**Comment by rhendric** [comment]
I propose merging, with the following to be done in future PRs:

* Deprecate or remove `extraRegex`
* Find some place to document common-updater-scripts

**Comment by AndersonTorres** [comment]
A `README.MD` in the same directory should be sufficient, in case no one else knows what to do.

**Pull request merged**